src/theory/bv/int_blaster.cpp

```diff
@@ -0,0 +1,1133 @@
+     * and binarized all its children.
+     * Now we binarize the current node itself.
+     */
+    kind::Kind_t k = current.getKind();
+    if ((numChildren > 2)
+        && (k == kind::BITVECTOR_PLUS || k == kind::BITVECTOR_MULT
+            || k == kind::BITVECTOR_AND || k == kind::BITVECTOR_OR
+            || k == kind::BITVECTOR_XOR || k == kind::BITVECTOR_CONCAT))
+    {
+      // We only binarize bvadd, bvmul, bvand, bvor, bvxor, bvconcat
+      Assert(d_binarizeCache.find(current[0]) != d_binarizeCache.end());
+      Node result = d_binarizeCache[current[0]];
+      for (uint64_t i = 1; i < numChildren; i++)
+      {
+        Assert(d_binarizeCache.find(current[i]) != d_binarizeCache.end());
+        Node child = d_binarizeCache[current[i]];
+        result = d_nm->mkNode(current.getKind(), result, child);
+      }
+      d_binarizeCache[current] = result;
+    }
+    else if (numChildren > 0)
+    {
+      // current has children, but we do not binarize it
+      NodeBuilder<> builder(k);
+      if (current.getMetaKind() == kind::metakind::PARAMETERIZED)
+      {
+        builder << current.getOperator();
+      }
+      for (Node child : current)
+      {
+        builder << d_binarizeCache[child].get();
+      }
+      d_binarizeCache[current] = builder.constructNode();
+    }
+    else
+    {
+      // current has no children
+      d_binarizeCache[current] = current;
+    }
+  }
+  return d_binarizeCache[n];
+}
+
+/**
+ * We traverse n and perform rewrites both on the way down and on the way up.
+ * On the way down we rewrite the node but not it's children.
+ * On the way up, we update the node's children to the rewritten ones.
+ * For each sub-node, we perform rewrites to eliminate operators.
+ * Then, the original children are added to toVisit stack so that we rewrite
+ * them as well.
+ */
+Node IntBlaster::eliminationPass(Node n)
+{
+  std::vector<Node> toVisit;
+  toVisit.push_back(n);
+  Node current;
+  while (!toVisit.empty())
+  {
+    current = toVisit.back();
+    // assert that the node is binarized
+    // The following variable is only used in assertions
+    CVC4_UNUSED kind::Kind_t k = current.getKind();
+    uint64_t numChildren = current.getNumChildren();
+    Assert((numChildren == 2)
+           || !(k == kind::BITVECTOR_PLUS || k == kind::BITVECTOR_MULT
+                || k == kind::BITVECTOR_AND || k == kind::BITVECTOR_OR
+                || k == kind::BITVECTOR_XOR || k == kind::BITVECTOR_CONCAT));
+    toVisit.pop_back();
+    bool inEliminationCache =
+        (d_eliminationCache.find(current) != d_eliminationCache.end());
+    bool inRebuildCache =
+        (d_rebuildCache.find(current) != d_rebuildCache.end());
+    if (!inEliminationCache)
+    {
+      // current is not the elimination of any previously-visited node
+      // current hasn't been eliminated yet.
+      // eliminate operators from it using rewrite rules
+      Node currentEliminated =
```

Comment:
I'm not a huge fan of this, since this will need to be refactored if/when BV rewriter is refactored.

I assume you are doing this because the default rewriter of BV won't suffice?

Reply:
Yes, for example, we eliminate Or (using the relatively new rewrite rule OrEliminate), which the default rewriter does not do. In general, it is expected that int blasting would prefer other rewrites than bit-blasting.

Reply:
Wouldn't it be simpler to just add the bv2int case for BVOR?

Reply:
And for the rest of these eliminations as well? 
Some of them were written specifically to be used in this pass, e.g.:
```
                                  RewriteRule<SdivEliminateFewerBitwiseOps>,
                                  RewriteRule<SremEliminateFewerBitwiseOps>,
                                  RewriteRule<SmodEliminateFewerBitwiseOps>,
```

Some of them are not turned on by the default rewriter.

We could have all these cases in the main translation, of coruse, but this feels like a rewrite doesn't it?

Reply:
The rest of the eliminations should be made available if we had access to Theory::expandDefinitions here. 

I have been advocating for moving Theory::expandDefinitions to Rewriter::expandDefinitions.

What else would remain if you were able to call `Rewriter::rewrite( Theory/Rewriter::expandDefinitions( n ) )`?

Reply:
Btw, I guess you can call `TheoryBVRewriter::eliminateBVSDiv(node);`.

Reply:
In what follows, whenever I say "not called by the bv rewriter" I actually mean "not called from this file: https://github.com/CVC4/CVC4/blob/master/src/theory/bv/theory_bv_rewriter.cpp". If these statements are not equivalent please tell me, i think they are.

Let's start with sdiv, srem, smod:
```
 RewriteRule<SdivEliminateFewerBitwiseOps>
 RewriteRule<SremEliminateFewerBitwiseOps>
 RewriteRule<SmodEliminateFewerBitwiseOps>
```
I don't think it would be good to call `TheoryBVRewriter::eliminateBVSDiv(node)`. It calls the ordinary elimination rules for sdiv, srem and smod. These introduce extraction terms in order to check the sign of the bit-vector. Extraction terms are translated to integers using divs and mods which is expensive. In contrast, I implemented custom sdiv, srem and smod elimination rules called `SdivEliminateFewerBitwiseOps,SremEliminateFewerBitwiseOps,SmodEliminateFewerBitwiseOps` that check the sign of the translated vector simply using the integer `<` operator. I think this approach is better for our translation.
I think this answer is also relevant for your question about `expandDefinition` because i seems like the usual eliminations are called from there as well.

The following elimination rules are not called from the bv rewriter because these operators are not eliminated by it. I added them for this pass, and hence calling the rewriter/expandDef won't help.
```
 RewriteRule<NegEliminate>
 RewriteRule<XorEliminate>
 RewriteRule<OrEliminate>
```

The following elimination rules are not called from the bv rewriter, I don't know why. Hence calling the rewriter won't help.
```
 RewriteRule<CompEliminate>
 RewriteRule<SltEliminate>
```

Now, the following rewrites are in fact being called from the BV rewriter. Does that mean that If I call `Rewriter::rewrite(n)` they will be invoked? If so, I can at least remove all these lines and replace them with assertions that these operators are absent from the processed formula.
```
RewriteRule<UdivZero>
 RewriteRule<XnorEliminate>
 RewriteRule<NandEliminate>
 RewriteRule<NorEliminate>
 RewriteRule<SubEliminate>
 RewriteRule<RepeatEliminate>
 RewriteRule<RotateRightEliminate>
 RewriteRule<RotateLeftEliminate>
 RewriteRule<SleEliminate>
 RewriteRule<SgtEliminate>
 RewriteRule<SgeEliminate>
```

Reply:
Yes, those that are normally rewritten should be omitted from this list.  
I would conjecture that Comp and Slt *are* eliminated, but through other means.

This means that Neg / Xor / Or are the only cases. I would prefer if your pass explicitly handled these as part of its conversion. I believe this would be more informative, instead of having to have an explicit dependency on external rewrite rules.

Reply:
Btw, the same applies for Sdiv / Smod / Srem, can you actually verify that your code actually has the opportunity to rewrite these?  The problem is that the standard rewriter can definitely be called before you run this pass, and the rewriter will eliminate all these if I'm not mistaken.

Reply:
> Yes, those that are normally rewritten should be omitted from this list.
> I would conjecture that things like Comp are eliminated, but through other rewrite rules.
OK sure, I will remove them and add assertions.

> This means that Neg / Xor / Or are the only cases (maybe SLT as well). I would prefer if your pass explicitly handled these as part of its conversion. I believe this would be more informative, instead of having to have an explicit dependency on external rewrite rules.
I am happy to do so, but would like to verify that @4tXJ7f is alright with that also, as he advocated for using rewrites some time ago in an offline discussion. 

 

> @ajreynol ajreynol 2 hours ago Member
> Btw, the same applies for Sdiv / Smod / Srem, can you actually verify that your code actually has the opportunity to rewrite these? The problem is that the standard rewriter can definitely be called before you run this pass, and the rewriter will eliminate all these if I'm not mistaken.
You are right -- they are not called! I believe I did verify that they are called when I introduced them but since then we changed (more than once) the place of this preprocessing pass w.r.t. expanding definitions and rewritings. This would be another advantage of the approach that you are suggesting (handling everything in the pass).